lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -3251,46 +3251,52 @@ ParseResult FIRStmtParser::parseRWProbe(Value &result) {
-    if (targetType.hasUninferredReset())
+  auto *definingOp = target.getDefiningOp();
+
+  if (isa_and_nonnull<MemOp, CombMemOp, SeqMemOp, MemoryPortOp,
+                      MemoryDebugPortOp, MemoryPortAccessOp>(definingOp))
+    return emitError(startTok.getLoc(), "cannot probe memories or their ports");
+
+  // Use Forceable if necessary (reset).
+  if (targetType.hasUninferredReset()) {
+    if (!definingOp)
       return emitError(startTok.getLoc(),
                        "must have concrete reset type in type ")
              << targetType;
-    auto forceableType =
-        firrtl::detail::getForceableResultType(true, targetType);
-    if (!forceableType)
-      return emitError(startTok.getLoc(), "cannot force target of type ")
-             << targetType;
 
-    // Get InnerRef for target field.
-    auto mod = cast<FModuleOp>(arg.getOwner()->getParentOp());
-    auto sym = getInnerRefTo(
-        hw::InnerSymTarget(arg.getArgNumber(), mod, fieldRef.getFieldID()),
-        [&](auto _) -> hw::InnerSymbolNamespace & { return modNameSpace; });
-    result = builder.create<RWProbeOp>(forceableType, sym);
-    return success();
-  }
+    auto forceable = dyn_cast<Forceable>(definingOp);
+    if (!forceable || !forceable.isForceable() /* e.g., is/has const type*/)
+      return emitError(startTok.getLoc(), "rwprobe target not forceable")
+          .attachNote(definingOp->getLoc());
 
-  auto *definingOp = target.getDefiningOp();
-  if (!definingOp)
-    return emitError(startTok.getLoc(),
-                     "rwprobe value must be defined by an operation");
+    // TODO: do the ref.sub work while parsing the static expression.
+    result = getValueByFieldID(builder, forceable.getDataRef(),
+                               fieldRef.getFieldID());
 
-  if (isa<MemOp, CombMemOp, SeqMemOp, MemoryPortOp, MemoryDebugPortOp,
-          MemoryPortAccessOp>(definingOp))
-    return emitError(startTok.getLoc(), "cannot probe memories or their ports");
+    return success();
+  }
 
-  auto forceable = dyn_cast<Forceable>(definingOp);
-  if (!forceable || !forceable.isForceable() /* e.g., is/has const type*/)
-    return emitError(startTok.getLoc(), "rwprobe target not forceable")
-        .attachNote(definingOp->getLoc());
+  // RWProbe op!
+  auto forceableType = firrtl::detail::getForceableResultType(true, targetType);
+  if (!forceableType)
+    return emitError(startTok.getLoc(), "cannot force target of type ")
+           << targetType;
 
-  // TODO: do the ref.sub work while parsing the static expression.
-  result =
-      getValueByFieldID(builder, forceable.getDataRef(), fieldRef.getFieldID());
+  hw::InnerSymTarget innerSymTarget;
+  if (auto arg = dyn_cast<BlockArgument>(target)) {
+    auto mod = cast<FModuleOp>(arg.getOwner()->getParentOp());
+    innerSymTarget =
+        hw::InnerSymTarget(arg.getArgNumber(), mod, fieldRef.getFieldID());
+  } else {
+    innerSymTarget = hw::InnerSymTarget(definingOp, fieldRef.getFieldID());
+  }
```

Comment:
This feels like a nice utility.

Reply:
Totally! I'll pull this out, thanks!

Reply:
(no good way to do this generically across HW/FIRRTL without having dependencies on both, will find a place for it... :+1: )